keps/sig-api-machinery/1027-api-unions/README.md

```diff
@@ -86,42 +123,194 @@ become possible.
+  know about is currently set, the server can simply know to always respect the
+  discriminator. Without a discriminator, the server will have to do convoluted
+  logic to detect that the previously set field has not been modified and that
+  only one of the other union fields has been.
+
+### Go Markers
 
 We're proposing a new type of tags for go types (in-tree types, and also
 kubebuilder types):
 
-- `// +union` before a structure means that the structure is a union. All the
-  fields must be optional (beside the discriminator) and will be included as
-  members of the union. That structure CAN be embedded in another structure.
-- `// +unionDeprecated` before a field means that it is part of the
-  union. Multiple fields can have this prefix. These fields MUST BE optional,
-  omitempty and pointers. The field is named deprecated because we don't want
-  people to embed their unions directly in structures, and only exist because of
-  some existing core types (e.g. `Value` and `ValueFrom` in
-  [EnvVar](https://github.com/kubernetes/kubernetes/blob/3ebb8ddd8a21b/staging/src/k8s.io/api/core/v1/types.go#L1817-L1836)).
+
 - `// +unionDiscriminator` before a field means that this field is the
-  discriminator for the union. Only one field per structure can have this
-  prefix. This field HAS TO be a string, and CAN be optional.
+  discriminator for the union. This field MUST be an enum defined as a string (see section on
+  discriminator values). This field MUST be
+  required.
+- `// +unionMember=<memberName>,discriminatedBy=<discriminatorName>` before a field means that this
+  field is a member of a union. The `<memberName>` is the name of the field that will be set as the discriminator value.
+  It MUST correspond to one of the valid enum values of the discriminator's enum
+  type. It defaults to the `CamelCase` representation of the field name if not specified. The `<discriminatorName>` is optional if there
+  is only union in a struct and required if there are multiple unions per
+  struct. It should be the `CamelCase` representation of the field tagged with
+  `unionDiscriminator`.
+
+#### Discriminator Values
+
+Here we present a description of how discriminators and their valid values
+should be defined.
+
+As described above, the discriminator field must be a string and required.
+Because, there are only a few specific values that the discriminator can be, we
+propose that all discriminators should be defined as an enum, and should be
+tagged so via the enum go marker `// +enum`.
+
+Required unions will have the number of valid discriminator values equal to the
+number of member fields (see exception below on empty union members). Optional
+unions will have the number of valid discriminator values equal to the number of
+member fields, plus one additional value for when "no member" is desired. By
+convention, this "no member" discriminator value should be the empty string.
```

Comment:
Is it possible to have multiple "no member" options? For example I had a union previously where there were multiple options for the discriminator that currently required no configuration, but maybe did in the future. For example, what if the [DeploymentStrategy](https://github.com/kubernetes/api/blob/edebc6706ddef7e828b5b606fbd0c91c6843c2fa/apps/v1/types.go#L397) had more values for the strategy type? The StrategyType has a semantic value as a discriminant but maybe only `RollingUpdate` actually requires configuration. Does this count as an optional union?

Is this just covered by the empty union members case below?